<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An introduction to the standR package • standR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="An introduction to the standR package">
<meta property="og:description" content="standR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">standR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/standR_introduction.html">An introduction to the standR package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>An introduction to the standR package</h1>
                        <h4 data-toc-skip class="author">Ning Liu,
Dharmesh Bhuva, Ahmed Mohamed, Chin Wee Tan, Melissa Davis</h4>
            
            <h4 data-toc-skip class="date">2022-03-17</h4>
      
      
      <div class="hidden name"><code>standR_introduction.Rmd</code></div>

    </div>

    
    
<style type="text/css">
  body{
  font-size: 14pt;
}
</style>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <code>standR</code> package is short for <strong>S</strong>patial
<strong>t</strong>ranscriptomics <strong>a</strong>nalyzes
a<strong>n</strong>d <strong>d</strong>ecoding in <strong>R</strong>, it
aims at providing good practice pipeline and useful functions for users
to analyze Nanostring’s GeoMx DSP data, including data with GeoMX
protein (~30-60 markers), CTA (cancer transcriptome atlas, ~1800 genes)
and WTA (whole transcriptome atlas, ~18000 genes) panels.</p>
<p>The data analysis workflow consists of the following steps:</p>
<div class="figure">
<img src="../reference/figures/workflow.jpg" title="Workflow Steps" alt=""><p class="caption">Workflow Steps</p>
</div>
<p>The purpose of this vignette is to demonstrate how to use standR to
analyze Nanostring GeoMx DSP data.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The development version of <code>standR</code> can be installed from
GitHub:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"DavisLaboratory/standR"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-packages-and-data">Load packages and data<a class="anchor" aria-label="anchor" href="#load-packages-and-data"></a>
</h2>
<p>We will load the <code>standR</code> package. We designed the package
and the analysis workflow to be consistent with the BioConductor’s
Orchestrating Spatially Resolved Transcriptomics Analysis (OSTA)
framework, namely the <code>SpatialExperiment</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">standR</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/limma" class="external-link">limma</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></code></pre></div>
<p>In this vignette, we will use a publicly available WTA dataset of
diabetic kidney disease (DKD) as the case study.</p>
<div class="section level3">
<h3 id="import-from-paths">Import from paths<a class="anchor" aria-label="anchor" href="#import-from-paths"></a>
</h3>
<p>Data we are using here is a published whole transcriptome atlas (WTA)
dataset of diabetic kidney disease (DKD).</p>
<p>The data is available <a href="http://nanostring-public-share.s3-website-us-west-2.amazonaws.com/GeoScriptHub/KidneyDataset/" class="external-link">here</a>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dirPath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">url_base</span> <span class="op">&lt;-</span> <span class="st">"http://nanostring-public-share.s3-website-us-west-2.amazonaws.com/GeoScriptHub/KidneyDataset/"</span>

<span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">url_base</span>,<span class="st">"Kidney_Raw_TargetCountMatrix.txt"</span><span class="op">)</span>, destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dirPath</span>,<span class="st">"Kidney_Raw_TargetCountMatrix.txt"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">url_base</span>,<span class="st">"Kidney_Sample_Annotations.txt"</span><span class="op">)</span>, destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dirPath</span>,<span class="st">"Kidney_Sample_Annotations.txt"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">url_base</span>,<span class="st">"Kidney_Feature_Annotations.txt"</span><span class="op">)</span>, destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dirPath</span>,<span class="st">"Kidney_Feature_Annotations.txt"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>This is the background for the data:</p>
<p>NanoString GeoMx DSP dataset of diabetic kidney disease (DKD) vs
healthy kidney tissue. <strong>Seven slides</strong> were analyzed,
<strong>4 DKD and 3 healthy</strong>. Regions of Interest (ROI) were
focused two different parts of a kidney’s structure: <strong>tubules or
glomeruli</strong>. Individual glomeruli were identified by a
pathologist as either <strong>relatively healthy or diseased</strong>
regardless if the tissue was DKD or healthy. Tubule ROIs were segmented
into <strong>distal (PanCK) and proximal (neg) tubules</strong>. While
both distal and proximal tubules are called tubules, they perform very
different functions in the kidney.</p>
<p>In order to import the data into a spatial experiment object. We use
the <code>readGeoMx</code> function.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">countFile</span> <span class="op">&lt;-</span> <span class="st">"Kidney_Raw_TargetCountMatrix.txt"</span>
<span class="va">sampleAnnoFile</span> <span class="op">&lt;-</span> <span class="st">"Kidney_Sample_Annotations.txt"</span>
<span class="va">featureAnnoFile</span> <span class="op">&lt;-</span> <span class="st">"Kidney_Feature_Annotations.txt"</span>

<span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readGeoMx.html">readGeoMx</a></span><span class="op">(</span><span class="va">dirPath</span>, <span class="va">countFile</span>, <span class="va">sampleAnnoFile</span>, featureAnnoFile <span class="op">=</span> <span class="va">featureAnnoFile</span>, hasNegProbe <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Next, we inspect the SpatialExperiment object. For more details, see
<a href="http://www.bioconductor.org/packages/release/bioc/html/SpatialExperiment.html" class="external-link">SpatialExperiment</a>.</p>
<p>We can see the dataset has 18503 genes and 231 ROIs (regions of
interest).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># check out object</span>
<span class="va">spe</span></code></pre></div>
<pre><code><span class="co">## class: SpatialExperiment </span>
<span class="co">## dim: 18503 231 </span>
<span class="co">## metadata(1): NegProbes</span>
<span class="co">## assays(2): counts logcounts</span>
<span class="co">## rownames(18503): PADI2 CYP24A1 ... FAM166A AGTPBP1</span>
<span class="co">## rowData names(3): RTS_ID ProbeID Negative</span>
<span class="co">## colnames(231): disease3_scan | 001 | PanCK disease3_scan | 001 | neg</span>
<span class="co">##   ... disease1B_scan | 023 | Geometric Segment disease1B_scan | 024 |</span>
<span class="co">##   Geometric Segment</span>
<span class="co">## colData names(25): SlideName ScanName ... RoiReportY sample_id</span>
<span class="co">## reducedDimNames(0):</span>
<span class="co">## mainExpName: NULL</span>
<span class="co">## altExpNames(0):</span>
<span class="co">## spatialCoords names(2) : ROICoordinateX ROICoordinateY</span>
<span class="co">## imgData names(0):</span></code></pre>
<p>Here we can see two types of count matrix, raw count in the
<code>count</code> assay and logCPM count in the <code>logcount</code>
assay.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">assayNames</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "counts"    "logcounts"</span></code></pre>
</div>
<div class="section level3">
<h3 id="import-from-dgelist-object">Import from DGEList object<a class="anchor" aria-label="anchor" href="#import-from-dgelist-object"></a>
</h3>
<p>Alternatively, <code>standR</code> provided a function to generate a
spatial experiment object from a DGEList object (<code>edgeR</code>
package).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dge</span> <span class="op">&lt;-</span> <span class="fu">edgeR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html" class="external-link">SE2DGEList</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span>

<span class="va">spe2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readGeoMxFromDGE.html">readGeoMxFromDGE</a></span><span class="op">(</span><span class="va">dge</span><span class="op">)</span>

<span class="va">spe2</span></code></pre></div>
<pre><code><span class="co">## class: SpatialExperiment </span>
<span class="co">## dim: 18503 231 </span>
<span class="co">## metadata(0):</span>
<span class="co">## assays(2): counts logcounts</span>
<span class="co">## rownames(18503): PADI2 CYP24A1 ... FAM166A AGTPBP1</span>
<span class="co">## rowData names(3): RTS_ID ProbeID Negative</span>
<span class="co">## colnames(231): disease3_scan | 001 | PanCK disease3_scan | 001 | neg</span>
<span class="co">##   ... disease1B_scan | 023 | Geometric Segment disease1B_scan | 024 |</span>
<span class="co">##   Geometric Segment</span>
<span class="co">## colData names(28): group lib.size ... RoiReportY sample_id</span>
<span class="co">## reducedDimNames(0):</span>
<span class="co">## mainExpName: NULL</span>
<span class="co">## altExpNames(0):</span>
<span class="co">## spatialCoords names(0) :</span>
<span class="co">## imgData names(0):</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<p>The quality control (QC) procedures are consisting of three major
steps:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Inspection of the sample metadata</strong>: sample
metadata can be seen as tabular-like format using the
<code>colData</code> function, however here we aim to visualise the
relations among important sample metadata, such as what slides did the
ROI come from, which are the control groups and treatment groups, what
are the pre-defined tissue types etc. By doing this step, we can have a
general idea of some important question such as how the experiment was
designed, should we be looking out for batch effect, and what interested
comparison can be established.</p></li>
<li><p><strong>Gene level QC</strong>: at gene level, by default we aim
at removing genes that are not expressed in more than 90% of the ROIs,
and identifying any ROIs with only few genes that are expressed. This is
the similar process as <code><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">edgeR::filterByExpr</a></code>, as genes with
consistently low counts are unlikely be identified as DE genes, and only
keeping genes with sufficiently large counts in the DE analysis can
increase statistical power while reduce multiple testing
burden.</p></li>
<li><p><strong>ROI level QC</strong>: at ROI level, we aim at identify
low-quality ROIs with relatively small library size and low cell count.
Problematic ROIs that are not removed could show up as separate clusters
in PCA/UMAPs and further affect the comparisons in DE analysis.</p></li>
</ol>
<div class="section level3">
<h3 id="visualise-metadata">Visualise metadata<a class="anchor" aria-label="anchor" href="#visualise-metadata"></a>
</h3>
<p>To visualise sample metadata, we can use the
<code>plotMetadata</code> function. In this dataset, slides, diseases
status, tissue regions, and different subtypes of the tissues are the
important features that we could input into the function.</p>
<p>We can see that all glomerulus are classified as abnormal and
healthy, and tubule are classified as neg and PanCK.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://corybrunson.github.io/ggalluvial/" class="external-link">ggalluvial</a></span><span class="op">)</span>

<span class="fu"><a href="../reference/plotMetadata.html">plotMetadata</a></span><span class="op">(</span><span class="va">spe</span>, column2plot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SlideName"</span>,<span class="st">"disease_status"</span>,<span class="st">"region"</span>,<span class="st">"pathology"</span>,
                                  <span class="st">"SegmentLabel"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>We therefore merge the region-related annotations to avoid
collinearity, which can affect the process of batch correction.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-colData.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-colData.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">region</span>,<span class="st">"_"</span>,<span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-colData.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">SegmentLabel</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_Geometric Segment"</span>,<span class="st">""</span>,<span class="va">.</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">.</span>,<span class="st">"_"</span>,<span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-colData.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">pathology</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_NA"</span>,<span class="st">"_ns"</span>,<span class="va">.</span><span class="op">)</span>


<span class="fu"><a href="../reference/plotMetadata.html">plotMetadata</a></span><span class="op">(</span><span class="va">spe</span>, column2plot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SlideName"</span>,<span class="st">"disease_status"</span>,<span class="st">"regions"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="gene-level-qc">Gene level QC<a class="anchor" aria-label="anchor" href="#gene-level-qc"></a>
</h3>
<p>Using the <code>addPerROIQC</code> function, we can add key
statistics to the <code>colData</code> of the object. In this case,
argument <code>rm_genes</code> is set to TRUE, with the default settings
of <code>min_count = 5</code> and <code>sample_fraction = 0.9</code>,
meaning here we first calculated a expression threshold on logCPM scale
(to account for library size differences), then genes that have
expression value below the threshold in more than 90% of the ROIs will
be removed. The count matrix with the removed gene are stored in the
<code>metadata</code> of the object, along with the calculated
expression threshold.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addPerROIQC.html">addPerROIQC</a></span><span class="op">(</span><span class="va">spe</span>, rm_genes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Here we can see that 121 genes are removed.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 18382   231</span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "NegProbes"         "lcpm_threshold"    "genes_rm_rawCount"</span>
<span class="co">## [4] "genes_rm_logCPM"</span></code></pre>
<p>Using the <code>plotGeneQC</code> function, we can have a look at
which were the genes removed and the overall distribution of percentage
of non-expressed genes in all ROIs. By default, top 9 genes are plotted
here (arranging by mean expression), user can increase the number of
plotted genes by changing the <code>top_n</code> parameter.</p>
<p>Moreover, users can order the samples (using parameter
<code>ordannots</code>) and color/shape the dots with annotation to find
out if these genes are specifically expressed in some samples (e.g. in
some tissue types or in some treatment group) so that we may need to
retain them.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotGeneQC.html">plotGeneQC</a></span><span class="op">(</span><span class="va">spe</span>, ordannots <span class="op">=</span> <span class="st">"regions"</span>, col <span class="op">=</span> <span class="va">regions</span>, point_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>In this case we don’t see any specific biological pattern for the
samples expressing this genes (figure above).</p>
</div>
<div class="section level3">
<h3 id="roi-level-qc">ROI level QC<a class="anchor" aria-label="anchor" href="#roi-level-qc"></a>
</h3>
<p>Using the <code>plotROIQC</code> function, we are able to visualise
some QC statistics in the ROI level, such as library size and cell count
(AOINucleiCount) (the default settinngs for this function).</p>
<p>In the ROI level QC, we first aim to identify (if any) ROI(s) that
have relatively low library size and low cell count because they are
considered as low quality samples due to insufficient sequencing depth
or lack of RNA in the chosen region.</p>
<p>In this case, looking at the distribution plots of library size and
nuclei count, we don’t see any particular spike in the low ends, rather
the distributions are relatively smooth. Looking at the dot plot,
library sizes are mostly positively correlate with the nuclei count,
with some data have relatively low library size while the nuclei count
is reasonable. We therefore can try to draw an filtering threshold at
the low end of the library size, in this case 50000. By coloring the dot
with their slide names, we find that the ROIs below the threshold are
all from slide disease1B, suggesting the reason for this might be some
technical issues of slide disease1B.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotROIQC.html">plotROIQC</a></span><span class="op">(</span><span class="va">spe</span>, y_threshold <span class="op">=</span> <span class="fl">50000</span>, col <span class="op">=</span> <span class="va">SlideName</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>Since library size of 50000 seems to be a reasonable threshold, here
we subset the spatial experiment object based on the library size in
<code>colData</code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spe</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-colData.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-colData.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">lib_size</span> <span class="op">&gt;</span> <span class="fl">50000</span><span class="op">]</span><span class="op">]</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="inspection-of-variations-on-roi-level">Inspection of variations on ROI level<a class="anchor" aria-label="anchor" href="#inspection-of-variations-on-roi-level"></a>
</h3>
<div class="section level4">
<h4 id="rle">RLE<a class="anchor" aria-label="anchor" href="#rle"></a>
</h4>
<p>We can use function <code>plotRLE</code> to visualise the relative
log expression (RLE) of the data to inspect the technical variation of
the data by looking at the distance from the median of the RLE (the
boxplot dot) to zero.</p>
<p>By default, we are plotting RLE of the raw count, where most of the
variation are from library size differences.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotRLE.html">plotRLE</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></code></pre></div>
<p>Using <code>assay = 2</code> to run RLE on the logCPM data, and we
can see that most technical variation from library sizes are removed.
And we can also sort the data using the annotation in the object by
specifying <code>ordannots</code>, with color or shape mapping parameter
as we usually write in ggplot, so that we can have a look at what factor
is heavily contributing to the technical variation.</p>
<p>Here we can see obvious variation from slides to slides, and small
variations are also observed within each slide.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotRLE.html">plotRLE</a></span><span class="op">(</span><span class="va">spe</span>, ordannots <span class="op">=</span> <span class="st">"SlideName"</span>, assay <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="va">SlideName</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h4>
<p>Using the <code>plotPCA</code> function, we can perform principal
component analysis (PCA) on the data. The PCA plot can help visualising
the variation (both biological and technical) in the data and finding
out which are the main factors contributing to the variations.</p>
<p>Here we color the PCA with slide information, and shape by regions
(tissue type). We can see that PC1 is mainly spread out by regions,
especially glomerulus and tubule. And grouping based on slide within
each tissue are observed. The subtypes in tubule are clearly separated,
but different subtypes of glomerulus is still grouping together.
Moreover, diseased tissues and control tissues are mixed as well
(disease slides and normal slides).</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotPCA.html">plotPCA</a></span><span class="op">(</span><span class="va">spe</span>, assay <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="va">SlideName</span>, shape <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="mds">MDS<a class="anchor" aria-label="anchor" href="#mds"></a>
</h4>
<p>Similarly, <code>plotMDS</code> can be used to visualise the
multidimension scaling of the data.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">standR</span><span class="fu">::</span><span class="fu"><a href="../reference/plotMDS.html">plotMDS</a></span><span class="op">(</span><span class="va">spe</span>, assay <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="va">SlideName</span>, shape <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="umap">UMAP<a class="anchor" aria-label="anchor" href="#umap"></a>
</h4>
<p>Further more, to be consistent with the <code>scater</code> package,
which is commonly used in single cell data and visium data analysis, we
provide the function <code>plotDR</code> to visualise any kinds of
dimension reduction results generated from <code>scater::run*</code>,
such as UMAP, TSNE and NMF, by specifying the <code>dimred</code>
parameter.</p>
<p>Here we plot the UMAP of our data. Similar variation can be observed
like PCA and MDS above.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotDR.html">plotDR</a></span><span class="op">(</span><span class="va">spe</span>, dimred <span class="op">=</span> <span class="st">"UMAP"</span>, col <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="other-pca-related-visualizations">Other PCA-related visualizations<a class="anchor" aria-label="anchor" href="#other-pca-related-visualizations"></a>
</h4>
<p>The <code>standR</code> package also provide other functions to
visualise the PCA, including the PCA scree plot, pair-dimension PCA plot
and PCA bi-plot.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotScreePCA.html">plotScreePCA</a></span><span class="op">(</span><span class="va">spe</span>, assay <span class="op">=</span> <span class="fl">2</span>, dims <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotPairPCA.html">plotPairPCA</a></span><span class="op">(</span><span class="va">spe</span>, col <span class="op">=</span> <span class="va">regions</span>, 
            shape <span class="op">=</span> <span class="va">disease_status</span>, assay <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotPCAbiplot.html">plotPCAbiplot</a></span><span class="op">(</span><span class="va">spe</span>, n_loadings <span class="op">=</span> <span class="fl">10</span>, assay <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h2>
<p>As we observed the technical variations in the data in both RLE and
PCA plots. It is necessary to perform normalization in the data.</p>
<p>In the <code>standR</code> package, we offer normalization options
including TMM, RPKM, TPM, CPM, upperquartile and sizefactor. Among them,
RPKM and TPM required gene length information (add
<code>genelength</code> column to the <code>rowData</code> of the
object). For TMM, upperquartile and sizefactor, their normalized factor
will be stored their <code>metadata</code>.</p>
<p>Here we used TMM to normalize the data.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-colData.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">biology</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-colData.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">disease_status</span>, <span class="st">"_"</span>, <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-colData.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">regions</span><span class="op">)</span>

<span class="va">spe_tmm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geomxNorm.html">geomxNorm</a></span><span class="op">(</span><span class="va">spe</span>, method <span class="op">=</span> <span class="st">"TMM"</span><span class="op">)</span></code></pre></div>
<p>Then we use RLE and PCA plot to assess the normalization count.</p>
<p>The RLE plot show most of the median of RLE are close to zero,
indicating that lots of technical variation are removed.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotRLE.html">plotRLE</a></span><span class="op">(</span><span class="va">spe_tmm</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">SlideName</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"TMM"</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<p>However, batch effect from the different slides are still observed,
confounding the separation between disease and normal.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotPairPCA.html">plotPairPCA</a></span><span class="op">(</span><span class="va">spe_tmm</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">disease_status</span>, shape <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="batch-correction">Batch correction<a class="anchor" aria-label="anchor" href="#batch-correction"></a>
</h2>
<p>In the Nanostring’s GeoMX DSP protocol, due to the fact that one
slide is only big enough for a handful of tissue segments (ROIs), it is
common that we see the DSP data being confounded by the batch effect
introduced by different slides. In order to establish fair comparison
between ROIs later on, it is necessary to remove this batch effect from
the data.</p>
<p>In the <code>standR</code> package, we provide two approaches of
removing batch effects, including RUV4 and Limma.</p>
<div class="section level3">
<h3 id="correction-method-ruv4">Correction method : RUV4<a class="anchor" aria-label="anchor" href="#correction-method-ruv4"></a>
</h3>
<p>To run RUV4 batch correction, we need to provide a list of “negative
control genes (NCGs)”.</p>
<p>The function <code>findNCGs</code> allows identifying the NCGs from
the data. In this case, since the batch effect is mostly introduced by
slide, we therefore want to identify NCGs across all slides, so here we
set the <code>batch_name</code> to “SlideName”, and select the top 500
least variable genes across different slides as NCGs.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findNCGs.html">findNCGs</a></span><span class="op">(</span><span class="va">spe</span>, batch_name <span class="op">=</span> <span class="st">"SlideName"</span>, top_n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span>

<span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "NegProbes"         "lcpm_threshold"    "genes_rm_rawCount"</span>
<span class="co">## [4] "genes_rm_logCPM"   "NCGs"</span></code></pre>
<p>Now we can run RUV4 on the data using function
<code>geomxBatchCorrection</code>. By default this function will use
<code>RUV4</code> to normalize the data. For RUV4 correction, the
function is requiring 3 parameters other than the input object,
including <code>factors</code>: the factor of interest, i.e. the
biological variation we plan to keep; <code>NCGs</code>: the list of
negative control genes detected using the function
<code>findNCGs</code>; and <code>k</code>: is the number of unwanted
factors to use, in the RUV documentation, it is suggest that we should
use the smallest k once we don’t observe technical variation in the
data.</p>
<p>Here we can find the best k by using function <code>findBestK</code>
to examine the silhouette score of ruv4-normalized count data clustering
according to batch from k = 1 to 10, where the score is closing to zero
means we are more likely to successfully remove the batch effect from
the data. In this case, k = 5 is the satisfied k.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/findBestK.html">findBestK</a></span><span class="op">(</span><span class="va">spe</span>, maxK <span class="op">=</span> <span class="fl">10</span>, factor_of_int <span class="op">=</span> <span class="st">"biology"</span>, NCGs <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">NCGs</span>, factor_batch <span class="op">=</span> <span class="st">"SlideName"</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
<p>We therefore will choose k=5 for this dataset.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spe_ruv</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geomxBatchCorrection.html">geomxBatchCorrection</a></span><span class="op">(</span><span class="va">spe</span>, factors <span class="op">=</span> <span class="st">"biology"</span>, 
                   NCGs <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">NCGs</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></code></pre></div>
<p>We can then inspect the PCA of the corrected data with annotations,
to inspect the removal of batch effects, and the retaining of the
biological factors.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotPairPCA.html">plotPairPCA</a></span><span class="op">(</span><span class="va">spe_ruv</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">disease_status</span>, shape <span class="op">=</span> <span class="va">regions</span>, title <span class="op">=</span> <span class="st">"RUV4"</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-30-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="correction-method-limma">Correction method: limma<a class="anchor" aria-label="anchor" href="#correction-method-limma"></a>
</h3>
<p>Another option is set the parameter <code>method</code> to “Limma”,
which uses the remove batch correction method from <code>limma</code>.
In this mode, the function is requiring 2 parameters, including
<code>batch</code>: a vector that indicating batches for all samples;
and <code>design</code>: a design matrix which is generated by
<code>model.matrix</code>, in the design matrix, all
biologically-relevant factors should be included.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spe_lrb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geomxBatchCorrection.html">geomxBatchCorrection</a></span><span class="op">(</span><span class="va">spe</span>,
                       batch <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-colData.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">SlideName</span>, method <span class="op">=</span> <span class="st">"Limma"</span>,
                       design <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">disease_status</span> <span class="op">+</span> <span class="va">regions</span>, 
                                             data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-colData.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Coefficients not estimable: batch6</span></code></pre>
<p>Again, using PCA to inspect the batch correction process.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotPairPCA.html">plotPairPCA</a></span><span class="op">(</span><span class="va">spe_lrb</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">disease_status</span>, shape <span class="op">=</span> <span class="va">regions</span>, title <span class="op">=</span> <span class="st">"Limma removeBatch"</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-32-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="evaluation">Evaluation<a class="anchor" aria-label="anchor" href="#evaluation"></a>
</h3>
<div class="section level4">
<h4 id="summary-statistics">Summary statistics<a class="anchor" aria-label="anchor" href="#summary-statistics"></a>
</h4>
<p>Besides looking at the PCA plots of normalized count, another way to
evaluate the batch correction is by summarising statistics such as
adjusted rand index, jaccard similarity coefficient, silhouette
coefficient and etc. We can do this by using the
<code>plotClusterEvalStats</code> function. In the biology section,
higher the score is better while in the batch section, lower is
better.</p>
<p>We can see that when it comes to stratifying for biology factors
(disease status and tissue regions) or measuring batch clustering for
this dataset, RUV4 outperform Limma in most statistics.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spe_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">spe</span>, <span class="va">spe_ruv</span>, <span class="va">spe_lrb</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotClusterEvalStats.html">plotClusterEvalStats</a></span><span class="op">(</span>spe_list <span class="op">=</span> <span class="va">spe_list</span>,
                     bio_feature_name <span class="op">=</span> <span class="st">"regions"</span>,
                     batch_feature_name <span class="op">=</span> <span class="st">"SlideName"</span>,
                     data_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Raw"</span>,<span class="st">"RUV4"</span>,<span class="st">"Limma"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-33-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="rle-plots">RLE plots<a class="anchor" aria-label="anchor" href="#rle-plots"></a>
</h4>
<p>Moreover, we can also have a look at the RLE plots of the normalized
count to determine which batch correction performs better for this
dataset.</p>
<p>We can clearly see that RUV4-corrected count have a overall
more-closer-to-zero median RLE compared to the limma-corrected data.
Therefore, in the downstream differential expression analysis, we would
suggest using the RUV4 as the batch correction method for this specific
dataset.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotRLE.html">plotRLE</a></span><span class="op">(</span><span class="va">spe_ruv</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">SlideName</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"RUV4"</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-34-1.png" width="700"></p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotRLE.html">plotRLE</a></span><span class="op">(</span><span class="va">spe_lrb</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">SlideName</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Limma removeBatch"</span><span class="op">)</span></code></pre></div>
<p><img src="standR_introduction_files/figure-html/unnamed-chunk-34-2.png" width="700"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R Under development (unstable) (2022-03-10 r81874)</span>
<span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">## Running under: Ubuntu 20.04.4 LTS</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">## [8] base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] ggalluvial_0.12.3           ggplot2_3.3.5              </span>
<span class="co">##  [3] magrittr_2.0.2              limma_3.51.5               </span>
<span class="co">##  [5] SpatialExperiment_1.5.4     SingleCellExperiment_1.17.2</span>
<span class="co">##  [7] SummarizedExperiment_1.25.3 Biobase_2.55.0             </span>
<span class="co">##  [9] GenomicRanges_1.47.6        GenomeInfoDb_1.31.5        </span>
<span class="co">## [11] IRanges_2.29.1              S4Vectors_0.33.11          </span>
<span class="co">## [13] BiocGenerics_0.41.2         MatrixGenerics_1.7.0       </span>
<span class="co">## [15] matrixStats_0.61.0          standR_0.99.0              </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] backports_1.4.1           systemfonts_1.0.4        </span>
<span class="co">##   [3] splines_4.2.0             BiocParallel_1.29.17     </span>
<span class="co">##   [5] scater_1.23.6             digest_0.6.29            </span>
<span class="co">##   [7] htmltools_0.5.2           viridis_0.6.2            </span>
<span class="co">##   [9] magick_2.7.3              fansi_1.0.2              </span>
<span class="co">##  [11] memoise_2.0.1             ScaledMatrix_1.3.0       </span>
<span class="co">##  [13] cluster_2.1.2             tzdb_0.2.0               </span>
<span class="co">##  [15] readr_2.1.2               R.utils_2.11.0           </span>
<span class="co">##  [17] vroom_1.5.7               pkgdown_2.0.2            </span>
<span class="co">##  [19] colorspace_2.0-3          ggrepel_0.9.1            </span>
<span class="co">##  [21] rbibutils_2.2.7           textshaping_0.3.6        </span>
<span class="co">##  [23] xfun_0.30                 dplyr_1.0.8              </span>
<span class="co">##  [25] crayon_1.5.0              RCurl_1.98-1.6           </span>
<span class="co">##  [27] jsonlite_1.8.0            glue_1.6.2               </span>
<span class="co">##  [29] ruv_0.9.7.1               gtable_0.3.0             </span>
<span class="co">##  [31] zlibbioc_1.41.0           XVector_0.35.0           </span>
<span class="co">##  [33] DelayedArray_0.21.2       car_3.0-12               </span>
<span class="co">##  [35] BiocSingular_1.11.0       DropletUtils_1.15.2      </span>
<span class="co">##  [37] Rhdf5lib_1.17.3           HDF5Array_1.23.2         </span>
<span class="co">##  [39] abind_1.4-5               scales_1.1.1             </span>
<span class="co">##  [41] DBI_1.1.2                 edgeR_3.37.0             </span>
<span class="co">##  [43] rstatix_0.7.0             Rcpp_1.0.8.2             </span>
<span class="co">##  [45] viridisLite_0.4.0         dqrng_0.3.0              </span>
<span class="co">##  [47] bit_4.0.4                 rsvd_1.0.5               </span>
<span class="co">##  [49] FNN_1.1.3                 ellipsis_0.3.2           </span>
<span class="co">##  [51] pkgconfig_2.0.3           R.methodsS3_1.8.1        </span>
<span class="co">##  [53] farver_2.1.0              scuttle_1.5.1            </span>
<span class="co">##  [55] sass_0.4.0                uwot_0.1.11              </span>
<span class="co">##  [57] locfit_1.5-9.5            utf8_1.2.2               </span>
<span class="co">##  [59] tidyselect_1.1.2          labeling_0.4.2           </span>
<span class="co">##  [61] rlang_1.0.2               munsell_0.5.0            </span>
<span class="co">##  [63] tools_4.2.0               cachem_1.0.6             </span>
<span class="co">##  [65] cli_3.2.0                 generics_0.1.2           </span>
<span class="co">##  [67] broom_0.7.12              evaluate_0.15            </span>
<span class="co">##  [69] stringr_1.4.0             fastmap_1.1.0            </span>
<span class="co">##  [71] mclustcomp_0.3.3          yaml_2.3.5               </span>
<span class="co">##  [73] ragg_1.2.2                knitr_1.37               </span>
<span class="co">##  [75] bit64_4.0.5               fs_1.5.2                 </span>
<span class="co">##  [77] purrr_0.3.4               nlme_3.1-155             </span>
<span class="co">##  [79] sparseMatrixStats_1.7.0   R.oo_1.24.0              </span>
<span class="co">##  [81] compiler_4.2.0            beeswarm_0.4.0           </span>
<span class="co">##  [83] ggsignif_0.6.3            tibble_3.1.6             </span>
<span class="co">##  [85] bslib_0.3.1               stringi_1.7.6            </span>
<span class="co">##  [87] highr_0.9                 desc_1.4.1               </span>
<span class="co">##  [89] RSpectra_0.16-0           lattice_0.20-45          </span>
<span class="co">##  [91] Matrix_1.4-0              vctrs_0.3.8              </span>
<span class="co">##  [93] pillar_1.7.0              lifecycle_1.0.1          </span>
<span class="co">##  [95] rhdf5filters_1.7.0        Rdpack_2.1.4             </span>
<span class="co">##  [97] jquerylib_0.1.4           BiocNeighbors_1.13.0     </span>
<span class="co">##  [99] cowplot_1.1.1             bitops_1.0-7             </span>
<span class="co">## [101] irlba_2.3.5               patchwork_1.1.1          </span>
<span class="co">## [103] R6_2.5.1                  gridExtra_2.3            </span>
<span class="co">## [105] vipor_0.4.5               assertthat_0.2.1         </span>
<span class="co">## [107] rhdf5_2.39.6              rprojroot_2.0.2          </span>
<span class="co">## [109] rjson_0.2.21              withr_2.5.0              </span>
<span class="co">## [111] GenomeInfoDbData_1.2.7    mgcv_1.8-39              </span>
<span class="co">## [113] parallel_4.2.0            hms_1.1.1                </span>
<span class="co">## [115] grid_4.2.0                beachmat_2.11.0          </span>
<span class="co">## [117] tidyr_1.2.0               rmarkdown_2.13           </span>
<span class="co">## [119] DelayedMatrixStats_1.17.0 carData_3.0-5            </span>
<span class="co">## [121] ggpubr_0.4.0              ggbeeswarm_0.6.0</span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ning Liu, Dharmesh D Bhuva, Ahmed Mohamed.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
